kv/bulk: stop ignoring errors from async flushes

rafiss opened the pull request:
###  kv/bulk: stop ignoring errors from async flushes

The error handling for Reset appears to have been broken by 50d101612ff. The error handling for Close never appears to have been correct.

fixes #143690
Release note (bug fix): Operations that write data in bulk, like index backfills or RESTORE jobs, previously could report incorrect errors when disk capacity was too low. This is fixed now.


### crosscluster: handle errors from sstbatcher.Close

This patch adds logic to swallow the error and log it. We may want to
take this further and propagate the error further up the call stack, but
I'm starting here for now.

rafiss committed:
kv/bulk: stop ignoring errors from async flushes

The error handling for Reset appears to have been broken by 50d101612ff.
The error handling for Close never appears to have been correct.

Release note (bug fix): Operations that write data in bulk, like index
backfills or RESTORE jobs, previously could report incorrect errors
when disk capacity was too low. This is fixed now.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144595)
<!-- Reviewable:end -->

rafiss committed:
crosscluster: handle errors from sstbatcher.Close

This patch adds logic to swallow the error and log it. We may want to
take this further and propagate the error further up the call stack, but
I'm starting here for now.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144595)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao, @msbutler, and @rytaft)

___
*[`pkg/crosscluster/logical/offline_initial_scan_processor.go` line 321 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144595#-OO4YuJp-w_DPE8fjYrG:-OO4YuJp-w_DPE8fjYrH:b-25yz4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/94b33b825d553db1563a4f6e625a127d2386b12d/pkg/crosscluster/logical/offline_initial_scan_processor.go#L321)):*
> ```Go
> 	if o.batcher != nil {
> 		if err := o.batcher.Close(o.Ctx()); err != nil {
> 			log.Errorf(o.Ctx(), "error on close(): %v", err)
> ```

for the DR folks: propagating the error here is going to take quite a bit more refactoring. i didn't do it yet since the call to workerGroup.Wait has the same issue. shall we do it in this PR, or would you want to take on that task separately?
___
*[`pkg/crosscluster/physical/stream_ingestion_processor.go` line 599 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144595#-OO4Z8BlB1t5KlQWjK3M:-OO4Z8BlB1t5KlQWjK3N:b-z3p5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/94b33b825d553db1563a4f6e625a127d2386b12d/pkg/crosscluster/physical/stream_ingestion_processor.go#L599)):*
> ```Go
> 	if sip.batcher != nil {
> 		if err := sip.batcher.Close(sip.Ctx()); err != nil {
> 			log.Errorf(sip.Ctx(), "error on close(): %v", err)
> ```

ditto above


<!-- Sent from Reviewable.io -->

jeffswenson commented:
I have somewhat strong opinions about error handling in the clean up path. I think it makes code overly complex and its easy to get wrong. The Flush() + Close() API works well where Flush() is the blocking API that returns errors to the users and Close() performs clean up in the unwind case. 

So I think the lack of error handling in `Reset` is actually okay, but we need to remove the wait in `Reset` and remove `Reset` from the public API because it is a foot gun. 

The reason https://github.com/cockroachdb/cockroach/pull/110218 was merged is it was using Reset in a place where we should have been closing the old batcher and creating a new batcher.

rafiss commented:
> I have somewhat strong opinions about error handling in the clean up path.

I'd be curious to learn more about the strong opinion. Error handling in the cleanup path is also part of [database/sql](https://pkg.go.dev/database/sql#Rows.Close) and our own internal executor interface, so this is already something we expect ourselves and our own customers to get right. 

It might be annoying, but I don't think that makes it an anti-pattern. In some cases, returning an error in `Close` makes a lot of sense and trying to avoid it could lead us to put the code through strange contortions. Anyways, as per the Reviewable comment I added to `(*offlineInitialScanProcessor).close`, handling errors in close already is something we might need to address elsewhere too.

---

As far as making `Reset` private, what do you think we should do about the usages in `(*offlineInitialScanProcessor).checkpoint`, `(*BufferingAdder).doFlush`, etc?

jeffswenson commented:
> I'd be curious to learn more about the strong opinion. Error handling in the cleanup path is also part of [database/sql](https://pkg.go.dev/database/sql#Rows.Close) and our own internal executor interface, so this is already something we expect ourselves and our own customers to get right.

A short version of my cleanup manifesto: Any API where we might want to handle an error in the close path should be structured as follows.

```
type Flushable interface {
   // Flush ensures that whatever side effect you care about is finished. If this is a file
   // Flush indicates that something went wrong and the task is not complete.
   Flush(context.Context) error
   // Close cleans up resources associated with the object. This means closing the file handle
   // or releasing the buffers. It may log errors, but there are no errors that can 
   // be meaningfully handled by the caller.
   Close(context.Context)
}
```

Using flushable would look something like:

```
func doThing(ctx context.Context, file string) error {
   flushable, err := OpenFile(file)
   if err != nil { return err}
   defer flushable.Close(ctx)
   
   // Do whatever you need. 

   return flushable.Flush(ctx)
}
```

I feel like go/sql is mostly an example of the flushable interface. If you consume `Rows` to completion, then the `Close(err)` isn't relevant. You actually need to check the rows error. Although that case is weird because if you need to exit early with an Okay result, the `Close` error becomes important. But that's bad design and not something to mimic.

> It might be annoying, but I don't think that makes it an anti-pattern.

I agree its common practice in the go community, but its also a common source of bugs. IMO designing APIs that put important errors in Close() is an anti-pattern because .Close() needs to be safe to defer() if you want code to be able to recover from panics. 


This slack thread contains an internal discussion on this from a few months ago: https://cockroachlabs.slack.com/archives/CLQ753JGH/p1726768924449369. I understand that not everyone agrees with me in general, but since this interface basically implements the Flush pattern already, I don't think we should make the Close errors load bearing.

> (*offlineInitialScanProcessor).close and Stream ingestion processor
If the processor is shutting down gracefully, it should call Flush and propagate an error if the Flush fails.

rafiss commented:
this has gone stale
